### PR TITLE
[Representation] Remove unknown templates from DOM

### PR DESCRIPTION
**Pull request by VWoeltjen**
Remove `mct-include` and `mct-representation` elements from the DOM when they do not contain a valid template; this supports different CSS approaches, per #195.

Summary of changes:
* Watch `key` for changes from `mct-include` (previously, this only displayed initial template choice, which has been sufficient for platform UI but would be unexpected.)
* Replace `mct-include` and `mct-representation` elements with comments whenever they have no template URL to populate themselves with, _or_ when the template URL is invalid
  * Remove usage of `ng-include`, instead loading and compiling templates via a service; this allows detection of cases where template URL is bad, and has useful side effect of reducing watch count slightly.
  * Replacing elements with comments was chosen over using `transclude: 'element'`, which may be used to get Angular to do the same. The problem with this approach is that Angular only allows element-level transclusion to occur on one directive per element; `mct-device`, `ng-if`, and several other directives use this and are commonly used in conjunction with `mct-include` and `mct-representation`
* Separate out the template-loading and comment-replacement behavior into a service used by both `mct-include` and `mct-representation`
* Add checks for the presence of `scope()` in `InfoGesture`; Angular sometimes returns `undefined` here at some times when representations are being transitioned back into the DOM
  * Unclear what the underlying cause of this is, as it appears to occur after elements have been reinserted into the DOM; believe Angular needs a digest cycle or similar after DOM changes before reattaching scope to element, but unsure how to verify this. However, [`scope()` cannot be relied upon anyway](https://github.com/angular/angular.js/issues/9515) so this check should be present in any event.

### CODE REVIEW

**Comment by VWoeltjen** [comment]
### Author Checklist

1. Changes address original issue? Y
2. Unit tests included and/or updated with changes? Y
3. Command line build passes? Y
4. Expect to pass code review? Y

File: platform/commonUI/inspect/src/gestures/InfoGesture.js [hunk 1]
```diff
@@ -131,6 +126,13 @@ define(
             this.pendingBubble = this.$timeout(displayBubble, this.delay);
 
             this.element.on('mouseleave', this.hideBubbleCallback);
+
+            // Also make sure we dismiss bubble if representation is destroyed
+            // before the mouse actually leaves it
+            this.scopeOff =
+                this.element.scope() &&
```
Comment (larkin) [inline platform/commonUI/inspect/src/gestures/InfoGesture.js]: This is to handle the case where mct-representation is removed from the DOM-- right? In this case, shouldn't the gesture service skip attaching gestures to invalid elements entirely?

Comment (VWoeltjen) [inline platform/commonUI/inspect/src/gestures/InfoGesture.js]: Representers (including GestureRepresenter) [do get skipped](https://github.com/nasa/openmctweb/blob/open195/platform/representation/src/MCTRepresentation.js#L182-L205) in these cases.

This handles a case where `scope()` returns undefined for reasons that are, well, unclear. The element has already re-entered the DOM at this point; I'm thinking that there's an interval between elements re-entering the DOM and being recognized by Angular (sufficient to provide answers to `scope()`) where this can occur.

I'm thinking this whole listener is actually unnecessary and should be removed. `MCTRepresentation` has the relevant scope already, is listening for `$destroy`, and is detaching all representers when this occurs.

File: platform/representation/src/MCTInclude.js [hunk 2]
```diff
@@ -54,36 +54,42 @@ define(
...
+                scope.$watch('key', function (key) {
+                    changeTemplate(key && templateMap[key]);
+                });
+            }
+
             // Prepopulate templateMap for easy look up by key
             templates.forEach(function (template) {
                 var key = template.key,
-                    path = $sce.trustAsResourceUrl([
+                    path = [
```
Comment (larkin) [inline platform/representation/src/MCTInclude.js]: Hrm, do you think getPath could be provided by the TemplateLinker?  Same code as the representation uses below.  ~~Can also standardize the optimizations: `mct-include` precalculates all paths, while `mct-representation` calculates paths on demand (and repeatedly); an in between solution could be memoizing calls to getPath using a WeakMap.~~

On second though; I think performance optimizations here are unnecessary.

File: platform/representation/src/TemplateLinker.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+            this.$log = $log;
+        }
+
+        TemplateLinker.prototype.load = function (templateUrl) {
+            var $http = this.$http,
+                $compile = this.$compile,
+                $log = this.$log,
+                templateMap = this.templateMap;
+
+            return (templateMap[templateUrl] = templateMap[templateUrl] ||
```
Comment (larkin) [inline platform/representation/src/TemplateLinker.js]: Mix of logic and control flow here; I can't tell what is actually returned here-- a promise for a compiled template, or an actual template?

What about something like:
```javascript
template = templateMap[templateUrl];
if (!template) {
    // make whatever type of object template is
}
return template;
```

File: platform/representation/src/TemplateLinker.js [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+         * @private
+         */
+        function TemplateLinker($http, $compile, $log) {
+            this.templateMap = {};
+            this.$http = $http;
+            this.$compile = $compile;
+            this.$log = $log;
+        }
+
+        TemplateLinker.prototype.load = function (templateUrl) {
```
Comment (larkin) [inline platform/representation/src/TemplateLinker.js]: Should this be marked `@private`?  Some JSDoc would help understand the purpose of this function.

File: platform/representation/src/TemplateLinker.js (hunk 3 above)
Comment (larkin) [inline platform/representation/src/TemplateLinker.js]: Ohh, on second thought can't this all be replaced with $templateRequest?  That handles requesting, compiling, and caching.

File: platform/representation/src/TemplateLinker.js [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+            function addElement() {
+                if (activeElement !== element) {
+                    activeElement.replaceWith(element);
+                    activeElement = element;
+                    activeElement.empty();
+                }
+            }
+
+            function populateElement(template) {
+                template(scope, function (innerClone) {
```
Comment (larkin) [inline platform/representation/src/TemplateLinker.js]: Hrm-- seems that this would be an ideal location to instantiate a new scope and destroy an old scope instead of reusing the same scope between different representations/templates/views.

**Comment by larkin** [comment]
@VWoeltjen this is the first time I'm really seeing the guts of 'representations' and 'views'.  I thought they were two distinctly different concepts, but upon seeing that views are considered representations, they seem to be two very similar things-- in terms of simplification, is it worth combining them?  (Happy to open a separate issue for this discussion)

**Comment by larkin** [comment]
@VWoeltjen angular provides a $templateRequest which I think you've recreated in the TemplateLinker-- I think you can reuse that to reduce the testing required on TemplateLinker.  Assigning back to you.

**Comment by VWoeltjen** [comment]
I've been thinking similarly w.r.t. `representation` and `views` - the only difference is that views are user-facing representations (e.g. they show up in the view switcher.) I think that's a change that can/should be considered in the context of #69; on the one hand, I'd like to add more flexibility around "views" in the MVC sense to make it easier to include non-Angular things, and on the other hand I'd like to reframe "views" in the MCT sense as "applications". The current paradigm means you start with a template, reference a controller, and build out your functionality from there. This works okay for a typical representation (e.g. a tree node), but for "views" this points you in the wrong direction - a view typically has a lot of internal functionality and state to manage, and more complex interactions with the information model. You're left either making architectural decisions about how to build each "view" (when those decisions could have been made once and reused), or not making those architectural decisions (because the platform API implies that this should be as easy as a template which references some code) and ending up with a mess.

**Comment by VWoeltjen** [comment]
I'll try to rewrite with `$templateRequest` (wasn't aware of this, makes much more sense.) Only downside is that templates aren't pre-compiled (so a separate cache or a new `$compile` is needed each time a template changes); but, `ng-include` does a new `$compile` each time so I guess I shouldn't be worrying about this.

**Comment by larkin** [comment]
It's views all the way down-- views are views are views; to an engineer, I don't think we should overload that term.

What do you think of switching the terms "representations" and "view",  such that:
* `views` are small reusable components-- building blocks of the application.
* `representations` are representations of domain objects that a user chooses between; they are a combination of a controller (which loads data) and a view(say a plot, or a table), plus some criteria specification that defines what they apply to (e.g. matching capabilities, model properties, type definitions, etc).

**Comment by VWoeltjen** [comment]
Agree, I'd rather have "view" mean something like "DOM elements + minimal wrapping to integrate with the application"; the difficulty is that we use the term "views" to describe a specific concept in the UI (stuff that appears in the right-hand area). But, we're not actually using that term _in_ the UI so we have some flexibility there, I suppose. 

There is a distinct category in between user-facing options for visualizing a domain object and general-purpose views (what's currently called a `representation`). I think we could make things simpler by having that middle category a simple subclass of views (e.g. a `DomainObjectView`, with capability matching etc. as appropriate) but I think that user-facing category should have a has-a relationship with views instead of an is-a relationship. There are a lot of recurring concepts in that user-facing category (transient internal state, persistent state, selection state to populate an external toolbar...) that are really not the concerns of a "view" in the MVC sense.

Anyway, this is already considered a bit in #69 (proposals "Arbitrary HTML Views", "Rename Views to Applications") so probably best to move the discussion there. Feel free to submit PR's to that branch to add proposals (or add to that branch directly, it's all just "proposals" at this point)

**Comment by VWoeltjen** [comment]
Summary of changes in response to code review:

* Removed scope usage from `InfoGesture` entirely
  * Verified that this does not have harmful side effects in cases this was meant to handle (e.g. when clicking on a grid item in a view of a folder - element is removed before `mouseleave` can fire, but info bubble still gets dismissed due to representation listening for `$destroy`
* Use `$templateRequest` from TemplateLinker to remove need to explicitly cache
* Moved conversion from extension definition to URL from `mct-include` and `mct-representation` into a method on TemplateLinker
* Added JSDoc tagged `@private` to `load` method

**Comment by VWoeltjen** [comment]
Reassigning to @larkin, believe this is ready to merge after handling feedback from review

File: platform/representation/src/TemplateLinker.js (hunk 5 above)
Comment (VWoeltjen) [inline platform/representation/src/TemplateLinker.js]: This is a good candidate location for this, but think this is out of scope (no pun intended) for this change. Need to make sure appropriate properties get exposed in each new scope that is created, and that scopes are appropriately synchronized with the scope of the `mct-include` or `mct-representation` itself (e.g. to pass in `parameters`), etc...

Comment (VWoeltjen) [inline platform/representation/src/TemplateLinker.js]: Added #227 for this change

File: platform/representation/src/TemplateLinker.js [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+        /**
+         * The `templateLinker` service is intended for internal use by
+         * the `mct-include` and `mct-representation` directives. It is
+         * used to support common behavior of directives; specifically,
+         * loading templates and inserting them into a specified element,
+         * and/or removing that element from the DOM when there is no
+         * template to populate it with.
+         *
+         * @param $http Angular's `$http` service
```
Comment (larkin) [inline platform/representation/src/TemplateLinker.js]: Doc string out of date with current parameters.

**Comment by larkin** [comment]
@VWoeltjen one small problem with docstrings but this otherwise looks good.  I'm going to attach a completed Review Checklist-- can you resolve that docstring and then integrate it?

# Review Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? Y
3. Code style and in-line documentation are appropriate? Y (provided small tweak to template linker constructor is made)
4. Commit messages meet standards? Y
5. Project-specific information isolated to appropriate branches? Y

**Comment by VWoeltjen** [comment]
Thanks for catching that - updated JSDoc, merged.
